test/dynamo/test_hooks.py

```diff
@@ -0,0 +1,401 @@
+# Owner(s): ["module: dynamo"]
+
+import functools
+
+import torch
+import torch._dynamo
+import torch._dynamo.test_case
+import torch._dynamo.testing
+from functorch.compile import nop
+from torch._functorch.aot_autograd import aot_module_simplified
+
+
+def global_hook_0(grad):
+    return grad * 4
+
+
+def global_hook_1(grad):
+    return grad / 2
+
+
+def global_hook_2(grad):
+    return grad * 3
+
+
+h0 = None
+
+
+class HooksTests(torch._dynamo.test_case.TestCase):
+    def test_tensor_only_register_hook_in_graph_lambda(self):
+        def fn(x):
+            x.register_hook(lambda grad: grad * 2)
+            return x
+
+        cnts = torch._dynamo.testing.CompileCounter()
+        fn = torch._dynamo.optimize(cnts)(fn)
+        v = torch.tensor([0.0, 0.0, 0.0], requires_grad=True)
+        v = fn(v)
+        v.backward(torch.tensor([1.0, 2.0, 3.0]))
+        self.assertEqual(v.grad, torch.tensor([2.0, 4.0, 6.0]))
+        self.assertEqual(cnts.frame_count, 0)
+
+    def test_tensor_register_hook_in_graph_lambda(self):
+        def fn(x, y, z):
+            x.register_hook(lambda grad: grad * 2)
+            return x, y * y, z * z
+
+        cnts = torch._dynamo.testing.CompileCounter()
+        fn = torch._dynamo.optimize(cnts)(fn)
+        v = torch.tensor([0.0, 0.0, 0.0], requires_grad=True)
+        v = fn(v, torch.randn([2, 2]), torch.randn([2, 2]))[0]
+        v.backward(torch.tensor([1.0, 2.0, 3.0]))
+        self.assertEqual(v.grad, torch.tensor([2.0, 4.0, 6.0]))
+        self.assertEqual(cnts.frame_count, 1)
+
+    def test_tensor_register_hook_in_graph_break_handle_lambda(self):
+        def fn(x, y, z):
+            handle = x.register_hook(lambda grad: grad * 2)
+            z = z * z
+            handle.remove()
+            x.register_hook(lambda grad: grad * 3)
+            return x, y * y, z
+
+        cnts = torch._dynamo.testing.CompileCounter()
+        fn = torch._dynamo.optimize(cnts)(fn)
+        v = torch.tensor([0.0, 0.0, 0.0], requires_grad=True)
+        v = fn(v, torch.randn([2, 2]), torch.randn([2, 2]))[0]
+        v.backward(torch.tensor([1.0, 2.0, 3.0]))
+        self.assertEqual(v.grad, torch.tensor([3.0, 6.0, 9.0]))
+        self.assertEqual(cnts.frame_count, 2)
+
+    def test_tensor_register_hook_multi_handle_return(self):
+        def fn(x, y, z):
+            handle = x.register_hook(lambda grad: grad * 2)
+            h2 = handle
```

Comment:
@ezyang this is one case you had questions about / covered w/r/t name preservation.